doc/design.md

```diff
@@ -0,0 +1,949 @@
+
+Examples of tags for data variables are:
+- `Data::Value`
+- `Data::Variance`
+- `Data::String`
+- `Data::Int`
+- `Data::ExperimentLog`. *Question: Should this actually be a coordinate?*
+
+Examples of tags coordinate variables are:
+- `Coord::Tof`
+- `Coord::dSpacing`
+- `Coord::Q`
+- `Coord::SpectrumNumber`
+- `Coord::DetectorId`
+- `Coord::SpectrumPosition`
+- `Coord::DetectorPosition`
+- `Coord::DetectorGrouping`
+- `Coord::RowLabel`
+- `Coord::X`, `Coord::Y`, `Coord::Z`
+- `Coord::Qx`, `Coord::Qy`, `Coord::Qz`
+- coordinates related to experiment environment parameters such as `Coord::Polarization` and `Coord::Temperature`
+- `Coord::Mask`
+
+The tag defines the underlying data type.
+
+We furthermore define:
+- Coordinate variables do not have a name and are thus identified only by their tag.
+  Since `Dataset` is a set this implies that there cannot be two coordinate variable with the same tag in the same dataset.
+  *Example: Having two variables `Coord::X` representing an X coordinate in a dataset is forbidden, and would indeed not make sense.*
+- Coordinate variables have an implicit unit that cannot be changed (at least apart from its scale) since it is implied by the tag defining the coordinate.
+  *Example: The X coordinate would always have a unit representing a *length*, e.g., `m` or `mm`.*
+- Data variables have a name in addition to their tag.
+  *Example: We can have a data variable holding neutron counts in `Data::Value` fom the "sample" run and a data variable holding counts for the "can" run in the *same* dataset.*
+
+The name of data variables is furthermore used to imply a grouping of variables.
+*Example: In a dataset containing data variables `Data::Value` and `Data::Variance` a matching variable name implies that the uncertainties stored in `Data::Variance` are associated with the data stored in `Data::Value`.*
+
+The data in `Variable` is held by a type-erased handle, i.e., a `Variable` can hold data of any type.
+The data handle also implements a copy-on-write mechanism, which makes `Variable` and `Dataset` cheap to copy and saves memory if only some variables in a dataset are modified.
+
+
+### <a name="overview-operations"></a>Operations
+
+Alongside the new `Dataset` type a series of basic and common operations will be provided.
+A mostly but not entirely complete list of operations to be supported is:
+1. Arithmetic operations: unary `-`, binary `+`, `-`, `*`, `/`
+1. `transpose`
+1. `slice` and `dice`
+1. `concatenate` and `merge`
+1. `sort`
+1. `rebin`
+1. `convert` (unit and/or dimension)
+1. `filter`
+1. Boolean/logical operations for mask operations
+1. `integrate`
+1. Basic statistics operations such as min, max, mean, and standard deviation
+1. Serialization and deserialization, mainly for saving to and loading from disk
+   *Question: Should disk I/O live in a separate module/library?*
+
```

Comment:
Probably sooner or later, but I wanted to avoid listing all of them. Added those later is no trouble since all unary operations are "the same" for implementation purposes.